# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
val housing = spark.read.format("csv")
                        .option("inferSchema", "true")
                        .option("header", "true")
                        .load("data/housing.csv")

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
housing.printSchema

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



Print the number of records in the dataset.

In [3]:
housing.count

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
housing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

Print the number of records with population more than 10000.

In [5]:
housing.where('population > 10000).count

23

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
housing.describe("housing_median_age", "total_rooms","median_house_value", "population").show

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
import org.apache.spark.sql.functions._

housing.agg(max("housing_median_age")).show

housing.agg(min("total_rooms")).show

housing.agg(avg("median_house_value")).show

+-----------------------+
|max(housing_median_age)|
+-----------------------+
|                   52.0|
+-----------------------+

+----------------+
|min(total_rooms)|
+----------------+
|             2.0|
+----------------+

+-----------------------+
|avg(median_house_value)|
+-----------------------+
|     206855.81690891474|
+-----------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
housing.groupBy("ocean_proximity").count.orderBy(desc("count")).show

+---------------+-----+
|ocean_proximity|count|
+---------------+-----+
|      <1H OCEAN| 9136|
|         INLAND| 6551|
|     NEAR OCEAN| 2658|
|       NEAR BAY| 2290|
|         ISLAND|    5|
+---------------+-----+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
housing.groupBy("ocean_proximity").agg(avg("median_house_value").alias("avg_value")).show

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
housing.createOrReplaceTempView("df")
spark.sql("SELECT ocean_proximity, avg(median_house_value) AS avg_value FROM df GROUP BY ocean_proximity").show

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
import org.apache.spark.ml.feature.VectorAssembler

val va = new VectorAssembler().setInputCols(Array("housing_median_age", "total_rooms", "median_house_value", "population")).setOutputCol("features")

val housingAttrs = va.transform(housing)

housingAttrs.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|            features|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|[41.0,880.0,45260...|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|[21.0,7099.0,3585...|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|[52.0,1467.0,3521...|
|  -122.25|   37.85|              52.0|     12

va = vecAssembler_5dd670437e87
housingAttrs = [longitude: double, latitude: double ... 9 more fields]


[longitude: double, latitude: double ... 9 more fields]

In [12]:
import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs, "features").head

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0


## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [13]:
val housingCol1 = housing.withColumn("rooms_per_household", 'total_rooms / 'households)
val housingCol2 = housingCol1.withColumn("bedrooms_per_room", 'total_bedrooms / 'total_rooms)
val housingExtra = housingCol2.withColumn("population_per_household", 'population / 'households)

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [14]:
val renamedHousing = housingExtra.withColumnRenamed("median_house_value", "label")
renamedHousing.printSchema

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- label: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)
 |-- rooms_per_household: double (nullable = true)
 |-- bedrooms_per_room: double (nullable = true)
 |-- population_per_household: double (nullable = true)



renamedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [15]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [16]:
for (c <- colNum) {
    print(s"Nulls in column $c: ")
    println(renamedHousing.where(col(c).isNull).count)
}

Nulls in column longitude: 0
Nulls in column latitude: 0
Nulls in column housing_median_age: 0
Nulls in column total_rooms: 0
Nulls in column total_bedrooms: 207
Nulls in column population: 0
Nulls in column households: 0
Nulls in column median_income: 0
Nulls in column rooms_per_household: 0
Nulls in column bedrooms_per_room: 207
Nulls in column population_per_household: 0


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [17]:
import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median").setInputCols(Array("total_bedrooms", "bedrooms_per_room")).setOutputCols(Array("total_bedrooms", "bedrooms_per_room"))
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

for (c <- colNum) {
    print(s"Nulls in column $c: ")
    println(imputedHousing.where(col(c).isNull).count)
}

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows

Nulls in column longitude: 0
Nulls in column latitude: 0
Nulls in column housing_median_age: 0
Nulls in column total_rooms: 0
Nulls in column total_bedrooms: 0
Nulls in column population: 0
Nulls in column households: 0
Nulls in column median_income: 0
Nulls in column rooms_per_household: 0
Nulls in column bedrooms_per_room: 0
Nulls in column population_per_household: 0


imputer = imputer_b19e1c6e94dc
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vector of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [18]:
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(colNum).setOutputCol("imputed_features")
val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler().setInputCol("imputed_features").setOutputCol("rescaled_features")
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|    imputed_features|   rescaled_features|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069...|
|  -122.22|   37.86|              21.0|     7099.0|        1

va = vecAssembler_6cf7cbad8fe2
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_b0e92ece3b04
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [19]:
renamedHousing.select("ocean_proximity").distinct.show

+---------------+
|ocean_proximity|
+---------------+
|         ISLAND|
|     NEAR OCEAN|
|       NEAR BAY|
|      <1H OCEAN|
|         INLAND|
+---------------+



### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [20]:
import org.apache.spark.ml.feature.StringIndexer

val indexer = new StringIndexer().setInputCol("ocean_proximity").setOutputCol("ocean_proximity_num")
val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_proximity_num|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|                3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0

indexer = strIdx_45edac1d5b6a
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [21]:
indexer.fit(renamedHousing).labels

Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [22]:
import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator().setInputCols(Array("ocean_proximity_num")).setOutputCols(Array("ocean_proximity_oh"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------------------+------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_proximity_num|ocean_proximity_oh|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------------------+------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|                3.0|     (4,[3],[1.0])|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    24

encoder = oneHotEncoder_a1628ccf3411
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [23]:
import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipeline = new Pipeline().setStages(Array(imputer, va, scaler))
val catPipeline = new Pipeline().setStages(Array(indexer, encoder))
val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline))
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+-------------------+------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|    imputed_features|   rescaled_features|ocean_proximity_num|ocean_proximity_oh|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+-------------------+------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.555

numPipeline = pipeline_d3329a868eb8
catPipeline = pipeline_7385ab6da7ea
pipeline = pipeline_3f47d1bff0d6
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [24]:
// val newCols = Array("longitude", "latitude", "housing_median_age", "total_rooms", "total_bedrooms", "population", "households", "median_income", "label", "rooms_per_household", "bedrooms_per_room", "population_per_household", "imputed_features", "rescaled_features", "ocean_proximity_num", "ocean_proximity_oh")
val newCols = Array("rescaled_features", "ocean_proximity_oh")
// val newCols = Array("rescaled_features")

val va2 = new VectorAssembler().setInputCols(newCols).setOutputCol("features")
val dataset = va2.transform(newHousing).select("features", "label")

dataset.show(5)

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-61.007269596069...|452600.0|
|[-61.002278409814...|358500.0|
|[-61.012260782324...|352100.0|
|[-61.017251968579...|341300.0|
|[-61.017251968579...|342200.0|
+--------------------+--------+
only showing top 5 rows



newCols = Array(rescaled_features, ocean_proximity_oh)
va2 = vecAssembler_876ad24f5af5
dataset = [features: vector, label: double]


[features: vector, label: double]

In [25]:
// dataset.select("rescaled_features").collect().take(2).foreach(println)

Name: Syntax Error.
Message: 
StackTrace: 

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [26]:
val Array(trainSet, testSet) = dataset.randomSplit(Array(80,20))

trainSet = [features: vector, label: double]
testSet = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [27]:
import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr = new LinearRegression
// lr.setFeaturesCol("features")
val lrModel = lr.fit(trainSet)
val trainingSummary = lrModel.summary

println(s"Coefficients: " + lrModel.coefficients + s", Intercept: " + lrModel.intercept)
println(s"RMSE: " + trainingSummary.rootMeanSquaredError)

Coefficients: [-54258.396940276856,-54406.69550294254,13855.809275936428,4922.564631256042,5503.361902724797,-47187.5986252858,41600.70813243934,79322.40043522371,6156.108600567483,16428.715111299534,453.39492673625057,-132577.3128881812,-167382.56277169552,-127364.47645919824,-135125.9416229], Intercept: -2255422.561708333
RMSE: 68087.54853727188


lr = linReg_8e9f67a0f584
lrModel = linReg_8e9f67a0f584
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@2bdbb58c


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@2bdbb58c

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [28]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 202944.9303200608| 94600.0|[-62.065401082150...|
|144958.89060419519|103600.0|[-62.040445150874...|
|180681.98006715858| 79000.0|[-62.025471592109...|
|209529.58680375665|106700.0|[-62.005506847089...|
| 210024.2161325491| 50800.0|[-62.005506847089...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 67163.19346229338


predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_84b45ab1a6bb
rmse = 67163.19346229338


67163.19346229338

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [29]:
import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor().setFeaturesCol("features").setLabelCol("label")

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|216554.98047858942| 94600.0|[-62.065401082150...|
|216554.98047858942|103600.0|[-62.040445150874...|
|151186.61417322836| 79000.0|[-62.025471592109...|
|216554.98047858942|106700.0|[-62.005506847089...|
| 181413.3696128561| 50800.0|[-62.005506847089...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 66786.72911940701


dt = dtr_7fe608952602
dtModel = DecisionTreeRegressionModel (uid=dtr_7fe608952602) of depth 5 with 63 nodes
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_b8c1f8d263a5
rmse = 66786.72911940701


66786.72911940701

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [30]:
import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor().setFeaturesCol("features").setLabelCol("label")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|195103.53682681845| 94600.0|[-62.065401082150...|
|158035.51035437398|103600.0|[-62.040445150874...|
|149392.76754775323| 79000.0|[-62.025471592109...|
|195183.66831088546|106700.0|[-62.005506847089...|
|202514.04900866814| 50800.0|[-62.005506847089...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 63216.60167939826


rf = rfr_5619d09923e3
rfModel = RandomForestRegressionModel (uid=rfr_5619d09923e3) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_8f6b9b72b2fe
rmse = 63216.60167939826


63216.60167939826

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [31]:
import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor().setFeaturesCol("features").setLabelCol("label")

// train the model
val gbModel = gb.fit(trainSet)

// make predictions on the test data
val predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|181797.98005483247| 94600.0|[-62.065401082150...|
|122016.71247122787|103600.0|[-62.040445150874...|
|  73719.4113173422| 79000.0|[-62.025471592109...|
|148977.33203534925|106700.0|[-62.005506847089...|
|162541.24898071645| 50800.0|[-62.005506847089...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 55970.12455264689


gb = gbtr_accfbeabf27e
gbModel = GBTRegressionModel (uid=gbtr_accfbeabf27e) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_117a0b9911f9
rmse = 55970.12455264689


55970.12455264689

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [32]:
import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder()
    .addGrid(rfModel.numTrees, Array(1,  5, 10))
    .addGrid(rfModel.maxDepth, Array(5, 10, 15))
    .build()

val evaluator = new RegressionEvaluator()
    .setLabelCol("label")
    .setPredictionCol("prediction")
    .setMetricName("rmse")

val cv = new CrossValidator()
    .setEstimator(rf)
    .setEvaluator(evaluator)
    .setEstimatorParamMaps(paramGrid)
    .setNumFolds(3)

val cvModel = cv.fit(dataset)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|102365.29411764706| 94600.0|[-62.065401082150...|
|106297.28679163972|103600.0|[-62.040445150874...|
| 81154.34001670845| 79000.0|[-62.025471592109...|
| 98857.72268907563|106700.0|[-62.005506847089...|
|           73210.0| 50800.0|[-62.005506847089...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 29132.427590099385


paramGrid = 


Array({
	rfr_5619d09923e3-maxDepth: 5,
	rfr_5619d09923e3-numTrees: 1
}, {
	rfr_5619d09923e3-maxDepth: 5,
	rfr_5619d09923e3-numTrees: 5
}, {
	rfr_5619d09923e3-maxDepth: 5,
	rfr_5619d09923e3-numTrees: 10
}, {
	rfr_5619d09923e3-maxDepth: 10,
	rfr_5619d09923e3-numTrees: 1
}, {
	rfr_5619d09923e3-maxDepth: 10,
	rfr_5619d09923e3-numTrees: 5
}, {
	rfr_5619d09923e3-maxDepth: 10,
	rfr_5619d09923e3-numTrees: 10
}, {
	rfr_5619d09923e3-maxDepth: 15,
	rfr_5619d09923e3-numTrees: 1
}, {
	rfr_5619d09923e3-maxDepth: 15,
	rfr_5619d09923e3-numTrees: 5
}, {
	rfr_5619d09923e3-maxDepth: 15,
	rfr_5619...


---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [33]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}

trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1", "The first input column")
    final val inputCol2 = new Param[String](this, "inputCol2", "The second input column")
    final val outputCol = new Param[String](this, "outputCol", "The output column")
    
  protected def validateAndTransformSchema(schema: StructType): StructType = {
    // Check that the input type is a double
    val idx1 = schema.fieldIndex($(inputCol1))
    val idx2 = schema.fieldIndex($(inputCol2))
    val field1 = schema.fields(idx1)
    val field2 = schema.fields(idx2)
    if (field1.dataType != DoubleType) {
        throw new Exception(s"Input1 type ${field1.dataType} did not match input type DoubleType")
    }
    if (field2.dataType != DoubleType) {
        throw new Exception(s"Input2 type ${field2.dataType} did not match input type DoubleType")
    }
    // Add the return field
    schema.add(StructField($(outputCol), DoubleType, false))
  }
}


defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [34]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.functions.{col, udf}

import org.apache.spark.sql.types.DoubleType
import org.apache.spark.sql.types.StructField

class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("configurablewordcount"))
    
    def setInputCol1(value: String): this.type = set(inputCol1, value)
    
    def setInputCol2(value: String): this.type = set(inputCol2, value)
    
    def setOutputCol(value: String): this.type = set(outputCol, value)

    override def copy(extra: ParamMap): MyTransformer = {
        defaultCopy(extra)
    }

    override def transformSchema(schema: StructType): StructType = {
        // Add the return field
        schema.add(StructField($(outputCol), DoubleType, false))
      }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
        dataset.withColumn($(outputCol), col($(inputCol1)) / col($(inputCol2)))
  }
}

defined class MyTransformer


Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [35]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = configurablewordcount_c40c0eda75eb


myDataset: Unit = ()


configurablewordcount_c40c0eda75eb

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [36]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

defined class Instance
defined object Helper


In [37]:
class LR() extends Serializable {
    
    
// https://stackoverflow.com/questions/44450889/why-does-spark-shell-fail-to-load-a-file-with-class-with-rdd-imported
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}
    
    
    
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

defined class LR


In [38]:
abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends org.apache.spark.ml.PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: org.apache.spark.ml.linalg.Vector, val trainingError: Array[Double])
    extends MyLinearModel[org.apache.spark.ml.linalg.Vector, MyLinearModelImpl] {

  override def copy(extra: org.apache.spark.ml.param.ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: org.apache.spark.ml.linalg.Vector): Double = {
//     println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

defined class MyLinearModel
defined class MyLinearModelImpl


In [39]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._

import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

defined class MyLinearRegression
defined class MyLinearRegressionImpl


In [40]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet)
val predictions = model.transform(trainSet)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Training
+--------------------+--------+--------------------+
|          prediction|   label|            features|
+--------------------+--------+--------------------+
|-8.66269455361453...| 85800.0|[-62.040445150874...|
|-8.64697881989682...|111400.0|[-62.020480405854...|
|-8.64826971515409E37| 76100.0|[-62.015489219599...|
|-8.65689589500352...| 73200.0|[-62.005506847089...|
|-8.66412506699742...| 78300.0|[-62.000515660834...|
+--------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 8.269284361187093E37


lr = linReg_2cdf9d704f6d
model = linReg_2cdf9d704f6d
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_b1b48618e446
rmse = 8.269284361187093E37


8.269284361187093E37

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

Col `PAY_0` should be renamed to `PAY_1`.

In [41]:
val cc = spark.read.format("csv")
    .option("inferSchema", "true")
    .option("header", "true")
    .load("data/ccdefault.csv")
    .withColumnRenamed("PAY_0", "PAY_1")

cc.cache()

cc = [ID: int, LIMIT_BAL: int ... 23 more fields]


[ID: int, LIMIT_BAL: int ... 23 more fields]

In [42]:
cc.printSchema

root
 |-- ID: integer (nullable = true)
 |-- LIMIT_BAL: integer (nullable = true)
 |-- SEX: integer (nullable = true)
 |-- EDUCATION: integer (nullable = true)
 |-- MARRIAGE: integer (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- PAY_1: integer (nullable = true)
 |-- PAY_2: integer (nullable = true)
 |-- PAY_3: integer (nullable = true)
 |-- PAY_4: integer (nullable = true)
 |-- PAY_5: integer (nullable = true)
 |-- PAY_6: integer (nullable = true)
 |-- BILL_AMT1: integer (nullable = true)
 |-- BILL_AMT2: integer (nullable = true)
 |-- BILL_AMT3: integer (nullable = true)
 |-- BILL_AMT4: integer (nullable = true)
 |-- BILL_AMT5: integer (nullable = true)
 |-- BILL_AMT6: integer (nullable = true)
 |-- PAY_AMT1: integer (nullable = true)
 |-- PAY_AMT2: integer (nullable = true)
 |-- PAY_AMT3: integer (nullable = true)
 |-- PAY_AMT4: integer (nullable = true)
 |-- PAY_AMT5: integer (nullable = true)
 |-- PAY_AMT6: integer (nullable = true)
 |-- DEFAULT: integer (nullable = tru

In [43]:
cc.count

30000

In [44]:
cc.show

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_1|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|     3272|     3455|     3261|       0|    1000|    1000|    1000|       0|    2000|    

In [45]:
cc.describe("LIMIT_BAL", "AGE", "PAY_1", "BILL_AMT1", "PAY_AMT1","DEFAULT").show
// cc.describe(cc.columns:_*).show

+-------+------------------+-----------------+------------------+-----------------+-----------------+-------------------+
|summary|         LIMIT_BAL|              AGE|             PAY_1|        BILL_AMT1|         PAY_AMT1|            DEFAULT|
+-------+------------------+-----------------+------------------+-----------------+-----------------+-------------------+
|  count|             30000|            30000|             30000|            30000|            30000|              30000|
|   mean|167484.32266666667|          35.4855|           -0.0167|       51223.3309|        5663.5805|             0.2212|
| stddev|129747.66156720246|9.217904068090155|1.1238015279973335|73635.86057552966|16563.28035402577|0.41506180569093254|
|    min|             10000|               21|                -2|          -165580|                0|                  0|
|    max|           1000000|               79|                 8|           964511|           873552|                  1|
+-------+---------------

In [46]:
for (c <- cc.columns) {
    print(s"Number of unique values in column $c: ")
//     cc.select(col(c)).distinct.show
    println(cc.select(col(c)).distinct.count)
}

Number of unique values in column ID: 30000
Number of unique values in column LIMIT_BAL: 81
Number of unique values in column SEX: 2
Number of unique values in column EDUCATION: 7
Number of unique values in column MARRIAGE: 4
Number of unique values in column AGE: 56
Number of unique values in column PAY_1: 11
Number of unique values in column PAY_2: 11
Number of unique values in column PAY_3: 11
Number of unique values in column PAY_4: 11
Number of unique values in column PAY_5: 10
Number of unique values in column PAY_6: 10
Number of unique values in column BILL_AMT1: 22723
Number of unique values in column BILL_AMT2: 22346
Number of unique values in column BILL_AMT3: 22026
Number of unique values in column BILL_AMT4: 21548
Number of unique values in column BILL_AMT5: 21010
Number of unique values in column BILL_AMT6: 20604
Number of unique values in column PAY_AMT1: 7943
Number of unique values in column PAY_AMT2: 7899
Number of unique values in column PAY_AMT3: 7518
Number of uniqu

In [47]:
for (c <- cc.columns) {
    print(s"Nulls in column $c: ")
    println(cc.where(col(c).isNull).count)
}

Nulls in column ID: 0
Nulls in column LIMIT_BAL: 0
Nulls in column SEX: 0
Nulls in column EDUCATION: 0
Nulls in column MARRIAGE: 0
Nulls in column AGE: 0
Nulls in column PAY_1: 0
Nulls in column PAY_2: 0
Nulls in column PAY_3: 0
Nulls in column PAY_4: 0
Nulls in column PAY_5: 0
Nulls in column PAY_6: 0
Nulls in column BILL_AMT1: 0
Nulls in column BILL_AMT2: 0
Nulls in column BILL_AMT3: 0
Nulls in column BILL_AMT4: 0
Nulls in column BILL_AMT5: 0
Nulls in column BILL_AMT6: 0
Nulls in column PAY_AMT1: 0
Nulls in column PAY_AMT2: 0
Nulls in column PAY_AMT3: 0
Nulls in column PAY_AMT4: 0
Nulls in column PAY_AMT5: 0
Nulls in column PAY_AMT6: 0
Nulls in column DEFAULT: 0


Take a look at class balance

In [48]:
cc.groupBy('DEFAULT).count.show

+-------+-----+
|DEFAULT|count|
+-------+-----+
|      1| 6636|
|      0|23364|
+-------+-----+



Classes are not balanced, but not extremely. For the scope of this assignment we prefer to avoid subsampling techniques or similar.

### Prepare one hot encoding

After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute ocean_proximity.

In [49]:
for (c <- cc.columns) {
    print(s"Number of unique values in column $c: ")
//     cc.select(col(c)).distinct.show
    println(cc.select(col(c)).distinct.count)
}

Number of unique values in column ID: 30000
Number of unique values in column LIMIT_BAL: 81
Number of unique values in column SEX: 2
Number of unique values in column EDUCATION: 7
Number of unique values in column MARRIAGE: 4
Number of unique values in column AGE: 56
Number of unique values in column PAY_1: 11
Number of unique values in column PAY_2: 11
Number of unique values in column PAY_3: 11
Number of unique values in column PAY_4: 11
Number of unique values in column PAY_5: 10
Number of unique values in column PAY_6: 10
Number of unique values in column BILL_AMT1: 22723
Number of unique values in column BILL_AMT2: 22346
Number of unique values in column BILL_AMT3: 22026
Number of unique values in column BILL_AMT4: 21548
Number of unique values in column BILL_AMT5: 21010
Number of unique values in column BILL_AMT6: 20604
Number of unique values in column PAY_AMT1: 7943
Number of unique values in column PAY_AMT2: 7899
Number of unique values in column PAY_AMT3: 7518
Number of uniqu

We can consider `SEX`, `EDUCATION` and `MARRIAGE` as categorical variables. While `PAY_#` columns still represent discrite but not categorical variable. Hence we esclude the latters from the one hot encoding step

In [50]:
// import org.apache.spark.ml.feature.IntegerIndexer

// val indexer = new IntegerIndexer()
//     .setInputCol("SEX")
//     .setOutputCol("SEX_ind")

// val ohHousing = indexer.fit(cc).transform(cc)

// ohHousing.show(5)

Name: Syntax Error.
Message: 
StackTrace: 

In [51]:
import org.apache.spark.ml.feature.OneHotEncoderEstimator

val hot_encoder = new OneHotEncoderEstimator().setInputCols(Array("SEX","EDUCATION","MARRIAGE")).setOutputCols(Array("SEX_oh","EDUCATION_oh","MARRIAGE_oh"))
val cc_oh = hot_encoder.fit(cc).transform(cc)

cc_oh.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+-------------+-------------+-------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_1|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|       SEX_oh| EDUCATION_oh|  MARRIAGE_oh|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+-------------+-------------+-------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|    (2,[],[])|(6,[2],[1.0])|(3,[1],[1.0])|
|  2|   120000|  2|        2

hot_encoder = oneHotEncoder_4dc0528f08ba
cc_oh = [ID: int, LIMIT_BAL: int ... 26 more fields]


[ID: int, LIMIT_BAL: int ... 26 more fields]

## Rescaling Continuous Vars

We need to create a feature vector with the numerical variables in order to fit the scaler.

In [52]:
val contCols = Array("AGE","PAY_1", "PAY_2", "PAY_3", "PAY_4", "PAY_5", "PAY_6", "BILL_AMT1", "BILL_AMT2", "BILL_AMT3", "BILL_AMT4", "BILL_AMT5", "BILL_AMT6", "PAY_AMT1", "PAY_AMT2", "PAY_AMT3", "PAY_AMT4", "PAY_AMT5", "PAY_AMT6")

contCols = Array(AGE, PAY_1, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)


Array(AGE, PAY_1, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)

In [53]:
val scalerVectorAssembler = new VectorAssembler().setInputCols(contCols).setOutputCol("to_scale_features")
val ccFeaturesForScaler = scalerVectorAssembler.transform(cc)

val ccScaler = new StandardScaler().setInputCol("to_scale_features").setOutputCol("rescaled_features")
val ccScaled = ccScaler.fit(ccFeaturesForScaler).transform(ccFeaturesForScaler)

ccScaled.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+--------------------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_1|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|   to_scale_features|   rescaled_features|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+--------------------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|(19,[0,1,2,3,4,5,...|(19,[0,1,2,3,4,5,...|
|  2|   120000|  2|        2

scalerVectorAssembler = vecAssembler_8171fecd866d
ccFeaturesForScaler = [ID: int, LIMIT_BAL: int ... 24 more fields]
ccScaler = stdScal_3b5511a3557a
ccScaled = [ID: int, LIMIT_BAL: int ... 25 more fields]


[ID: int, LIMIT_BAL: int ... 25 more fields]

## Pipelines

We now create one pipline for one hot encoding of the categorical values and scaling and we merge the results toghether with a final vector assembler to feed the model training

In [54]:
import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}
// val scalingPipeline = new Pipeline().setStages(Array(scalerVectorAssembler, ccScaler))
// val categoricalPipeline = new Pipeline().setStages(Array(hot_encoder))

val input_features = Array("rescaled_features", "SEX_oh", "EDUCATION_oh", "MARRIAGE_oh")
val finalVectorAssembler = new VectorAssembler().setInputCols(input_features).setOutputCol("features")
val ccPipeline = new Pipeline().setStages(Array(scalerVectorAssembler, ccScaler, hot_encoder, finalVectorAssembler))

// val ccPipeline = new Pipeline().setStages(Array(scalingPipeline, categoricalPipeline))
val newCC = ccPipeline.fit(cc).transform(cc).withColumnRenamed("DEFAULT","label")

newCC.show(1)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+---------+-------------+-------------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_1|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|   to_scale_features|   rescaled_features|   SEX_oh| EDUCATION_oh|  MARRIAGE_oh|            features|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+---------+-------------+-------------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|  

input_features = Array(rescaled_features, SEX_oh, EDUCATION_oh, MARRIAGE_oh)
finalVectorAssembler = vecAssembler_e043a0bf1684
ccPipeline = pipeline_ac57d269a94c
newCC = [ID: int, LIMIT_BAL: int ... 29 more fields]


[ID: int, LIMIT_BAL: int ... 29 more fields]

## Models

1. We are going to use cross validation and we are going to use it on the trainSet.

1. We will take the model with the best average performance (on the k-folds) based on the areaUnderROC. Hence model selection is done based on the train dataset.

1. We will then retrain the best model on all the train data and obtain the so-called **final model**.

1. Finally, we evaluate the performance of the model on the untouched test set to get a rough idea of it's performance with unseen data.

Refereces:
* https://machinelearningmastery.com/train-final-machine-learning-model/
* https://stats.stackexchange.com/questions/52274/how-to-choose-a-predictive-model-after-k-fold-cross-validation
* https://stats.stackexchange.com/questions/2306/feature-selection-for-final-model-when-performing-cross-validation-in-machine?rq=1

In [55]:
val Array(trainSet, testSet) = newCC.randomSplit(Array(80,20))
// trainSet.printSchema

trainSet = [ID: int, LIMIT_BAL: int ... 29 more fields]
testSet = [ID: int, LIMIT_BAL: int ... 29 more fields]


[ID: int, LIMIT_BAL: int ... 29 more fields]

### Logistic Regression

In [58]:
import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.tuning.CrossValidator
import org.apache.spark.ml.tuning.CrossValidatorModel
import org.apache.spark.ml.classification.{BinaryLogisticRegressionSummary, LogisticRegression}

val lr = new LogisticRegression

val lrParamGrid = new ParamGridBuilder()
    .addGrid(lr.regParam, Array(0.01, 0.1, 1))
    .addGrid(lr.elasticNetParam, Array(0, 0.2, 0.5, 0.8, 1) )
    .build()

val lrEvaluator = new BinaryClassificationEvaluator()
    .setLabelCol("label")

val lrCv = new CrossValidator()
    .setEstimator(lr)
    .setEvaluator(lrEvaluator)
    .setEstimatorParamMaps(lrParamGrid)
    .setNumFolds(5)

val lrCvModel = lrCv.fit(trainSet)

lr = logreg_84065f0847d4
lrParamGrid = 


Array({
	logreg_84065f0847d4-elasticNetParam: 0.0,
	logreg_84065f0847d4-regParam: 0.01
}, {
	logreg_84065f0847d4-elasticNetParam: 0.0,
	logreg_84065f0847d4-regParam: 0.1
}, {
	logreg_84065f0847d4-elasticNetParam: 0.0,
	logreg_84065f0847d4-regParam: 1.0
}, {
	logreg_84065f0847d4-elasticNetParam: 0.2,
	logreg_84065f0847d4-regParam: 0.01
}, {
	logreg_84065f0847d4-elasticNetParam: 0.2,
	logreg_84065f0847d4-regParam: 0.1
}, {
	logreg_84065...


In [59]:
lrCvModel.avgMetrics

Array(0.7223706724360789, 0.7138416244446336, 0.6987129449463176, 0.7214661615433376, 0.6994501450014804, 0.5, 0.7194628592915459, 0.7020156399447537, 0.5, 0.7167386492675234, 0.6907718900791809, 0.5, 0.7149469559110878, 0.6907718900791809, 0.5)

In [60]:
lrCvModel.avgMetrics.max

0.7223706724360789

In [61]:
import org.apache.spark.ml.tuning.CrossValidatorModel

// author: Adam Vogel
// source: https://stackoverflow.com/a/34667785/2496574
def getBestModelParameters(cvModel : CrossValidatorModel) : ParamMap = {
    cvModel.getEstimatorParamMaps
           .zip(cvModel.avgMetrics)
           .maxBy(_._2)
           ._1
}

getBestModelParameters: (cvModel: org.apache.spark.ml.tuning.CrossValidatorModel)org.apache.spark.ml.param.ParamMap


In [62]:
getBestModelParameters(lrCvModel)

{
	logreg_84065f0847d4-elasticNetParam: 0.0,
	logreg_84065f0847d4-regParam: 0.01
}


It appears that the best parameter has a L2 penalty: https://spark.apache.org/docs/2.2.0/api/scala/index.html#org.apache.spark.ml.classification.LogisticRegressionModel

### Decision Tree

In [73]:
import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.tuning.CrossValidator
import org.apache.spark.ml.classification.DecisionTreeClassifier

val dt = new DecisionTreeClassifier().setFeaturesCol("features").setLabelCol("label")


val dtParamGrid = new ParamGridBuilder()
    .addGrid(dt.maxDepth, Array(1, 2, 4, 8, 16, 30))
    .addGrid(dt.maxBins, Array(2, 4, 8, 16, 32))
//     .addGrid(dt.impurity, Array("entropy", "gini"))  // gini is the default and it is already taking a lot to train
    .build()

val dtEvaluator = new BinaryClassificationEvaluator()
    .setLabelCol("label")

val dtCv = new CrossValidator()
    .setEstimator(dt)
    .setEvaluator(dtEvaluator)
    .setEstimatorParamMaps(dtParamGrid)
    .setNumFolds(5)

val dtCvModel = dtCv.fit(trainSet)

dt = dtc_bb974abf03e8
dtParamGrid = 


Array({
	dtc_bb974abf03e8-maxBins: 2,
	dtc_bb974abf03e8-maxDepth: 1
}, {
	dtc_bb974abf03e8-maxBins: 4,
	dtc_bb974abf03e8-maxDepth: 1
}, {
	dtc_bb974abf03e8-maxBins: 8,
	dtc_bb974abf03e8-maxDepth: 1
}, {
	dtc_bb974abf03e8-maxBins: 16,
	dtc_bb974abf03e8-maxDepth: 1
}, {
	dtc_bb974abf03e8-maxBins: 32,
	dtc_bb974abf03e8-maxDepth: 1
}, {
	dtc_bb974abf03e8-maxBins: 2,
	dtc_bb974abf03e8-maxDepth: 2
}, {
	dtc_bb974abf03e8-maxBins:...


In [77]:
dtCvModel.avgMetrics

Array(0.5, 0.35514493085810095, 0.35514493085810095, 0.35514493085810095, 0.35514493085810095, 0.5, 0.355660164776286, 0.355660164776286, 0.355660164776286, 0.3556252349859368, 0.5, 0.3187754550963848, 0.31878017079518906, 0.318763104303972, 0.3186235641789642, 0.5115200020199855, 0.3876561210957487, 0.39236206095903964, 0.4345544486514704, 0.43985172000039174, 0.5149146552802102, 0.4485063863276876, 0.49043745902559077, 0.4983388218597103, 0.49237161915850314, 0.5024404392535866, 0.5832762529893649, 0.6182451598999844, 0.6318174383103992, 0.6235743368402186)

In [78]:
dtCvModel.avgMetrics.max

0.6318174383103992

In [79]:
getBestModelParameters(dtCvModel)

{
	dtc_bb974abf03e8-maxBins: 16,
	dtc_bb974abf03e8-maxDepth: 30
}


A max depth of 30 is kind of strange. It seems to be really deep.

### Random Forest

In [67]:
import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.tuning.CrossValidator
import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}

val rf = new RandomForestClassifier().setFeaturesCol("features").setLabelCol("label")

val rfParamGrid = new ParamGridBuilder()
    .addGrid(rf.maxDepth, Array(5,10,15))
    .addGrid(rf.numTrees, Array(20,50,100))
    .addGrid(rf.featureSubsetStrategy, Array("onethird", "sqrt", "log2"))
    .build()

val rfEvaluator = new BinaryClassificationEvaluator()
    .setLabelCol("label")  // default metric is "areaUnderROC"

val rfCv = new CrossValidator()
    .setEstimator(rf)
    .setEvaluator(rfEvaluator)
    .setEstimatorParamMaps(rfParamGrid)
    .setNumFolds(5)

val rfCvModel = rfCv.fit(trainSet)

rf = rfc_bceecdec9c9a
rfParamGrid = 


Array({
	rfc_bceecdec9c9a-featureSubsetStrategy: onethird,
	rfc_bceecdec9c9a-maxDepth: 5,
	rfc_bceecdec9c9a-numTrees: 20
}, {
	rfc_bceecdec9c9a-featureSubsetStrategy: sqrt,
	rfc_bceecdec9c9a-maxDepth: 5,
	rfc_bceecdec9c9a-numTrees: 20
}, {
	rfc_bceecdec9c9a-featureSubsetStrategy: log2,
	rfc_bceecdec9c9a-maxDepth: 5,
	rfc_bceecdec9c9a-numTrees: 20
}, {
	rfc_bceecdec9c9a-featureSubsetStrate...


In [68]:
rfCvModel.avgMetrics

Array(0.7658136888170546, 0.7677072017048785, 0.7657211429927058, 0.7685944043367656, 0.768477974329785, 0.7688016905805286, 0.769437671140867, 0.7695941458935551, 0.7694353266157593, 0.7697991601138611, 0.7701327304852985, 0.7689986187201514, 0.7711920218271059, 0.7715381409560738, 0.7723333166618109, 0.7730134696331267, 0.7726902498068904, 0.7737932462054553, 0.7558708874158344, 0.75994815063928, 0.7578202116472622, 0.7637149551939806, 0.7660193093482491, 0.7679949806896035, 0.7666083931947483, 0.7676079640330162, 0.7681445560083573)

In [69]:
rfCvModel.avgMetrics.max

0.7737932462054553

In [70]:
getBestModelParameters(rfCvModel)

{
	rfc_bceecdec9c9a-featureSubsetStrategy: log2,
	rfc_bceecdec9c9a-maxDepth: 10,
	rfc_bceecdec9c9a-numTrees: 100
}


In [83]:
rfCvModel.getEstimatorParamMaps(5)  // the parameters of the winner

{
	rfc_3d7f3034b832-maxBins: 4,
	rfc_3d7f3034b832-maxDepth: 4
}


Select the best model

## Model selection

We compare the best performance in terms of area under the ROC curve of the best performing model type

In [80]:
(lrCvModel.avgMetrics.max, dtCvModel.avgMetrics.max, rfCvModel.avgMetrics.max)

(0.7223706724360789,0.6318174383103992,0.7737932462054553)

The best model is random forest with the following hyperparameters

In [81]:
getBestModelParameters(rfCvModel)

{
	rfc_bceecdec9c9a-featureSubsetStrategy: log2,
	rfc_bceecdec9c9a-maxDepth: 10,
	rfc_bceecdec9c9a-numTrees: 100
}


### Strengts and Weaknesses

* **Decision Tree**: the best strength of the model is that it is failry simple conceptually and also the implementation is quite performant.
On the dowsides, the ROC area is the lowest one and the fact that the depth of the tree that obtains the highest average performance is 30, which to us seems quite deep
for a decisin tree
* **Logistic Regression**: the only weakness is that it compares worse than random forest on the train cross validation. The strength of this model is that it is very simple, both to understand and to implement.
* **Random Forest**: The strength is for sure that is the better performing model. The weakness is that with an esemble of 100 trees, it is basically a black box and it is very difficult to understand the behavior of the model with particular cases.

The decision on the model to pick ultimately depends on the application of the model and the costs associated to misclassification, and also if we want the model to provide an undestandable explanation of the decision taken.
In this case we prefer to use the model with the highest performance, regardless of the easy of interpretation it offers, which is random forest. The assumption is that banks do not have to provide any explanation on the decision to lend or not money, hence they are looking merely to minimize their losses.

On the other hand, we have based our decision only on the average performance on the 5 folds, but we have not taken into account the standard deviation (which is an hint on the variance of the model). Spark do not offer this value out-of-the box, but there are situations when a lowest standard deviation can be preferred to better average performance. Thus implying a certain confidence level while deploying the model in production.

## Training

We procede by training the random forest model on all the train data with the hyperparameters obtained from the cross validation

In [83]:
val rf = new RandomForestClassifier().setFeatureSubsetStrategy("log2").setMaxDepth(10).setNumTrees(100)
val rfModel = rf.fit(trainSet)

rf = rfc_0016cf3edecf
rfModel = RandomForestClassificationModel (uid=rfc_0016cf3edecf) with 100 trees


RandomForestClassificationModel (uid=rfc_0016cf3edecf) with 100 trees

## Performance on the test

Given that we do not have any new-data, we use the test set to estimate the ability of the model to classify unseen data.

The result is an indicative metrics that could not be used to take any decision concerning the model. If we would have been in a real situation we would have used all the data at hand to fit the model.

In [86]:
val testPredictions = rfModel.transform(testSet)
val finalEvaluator = new BinaryClassificationEvaluator()
finalEvaluator.evaluate(testPredictions)

testPredictions = [ID: int, LIMIT_BAL: int ... 32 more fields]
finalEvaluator = binEval_8f9a63306e5f


0.7714580263950247

The area under the ROC curve on the test set is really close to the result on the train set, which is an satisfying result.

## Improvements

1. A more in-depth analysis of the dataset should be performed. We didn't looked for correlation between the variables (including the target variable), neither we tried to perform feature engeneering.
1. We did not extracted the appropriate treshold for the binary classification. This can be done only by knowing the cost associated to false positives and false negatives. We can imagine that a false negatives (a person unable to pay back) is much more expensive than a false positive (not giving the credit, hence losing the interest rates). Depending on the disparity between the two cost of error we could set the $\beta$ parameter in the f-score, then calculate the value of tresholds that gives the highest f-score and then include it in the final model as the "deployment" option.
1. As said above, it would have been interesting to look also to the standard deviation of the perfornce of the classifiers on all the folds used in the cross-validation process.
1. We should have extracted the scaling parameters of the standard scaler only from the train dataset (hence it should be included in the cross validation pipeline), because by using all the dataset to do so we are contributing, in a way, to information leakage between the test and train set.
1. The classes (default or not) are unbalanced. Subsampling the not default cases could have given better results.
1. We did not checked that the class were balanced in the test set and that the distribution of each variables in the test set is compatible with the ones in the train set.


# Various stuff...

In [71]:
rfCv.explainParams

collectSubModels: whether to collect a list of sub-models trained during tuning. If set to false, then only the single best sub-model will be available after fitting. If set to true, then all sub-models will be available. Warning: For large models, collecting all sub-models can cause OOMs on the Spark driver (default: false)
estimator: estimator for selection (current: rfc_bceecdec9c9a)
estimatorParamMaps: param maps for the estimator (current: [Lorg.apache.spark.ml.param.ParamMap;@bb5423)
evaluator: evaluator used to select hyper-parameters that maximize the validated metric (current: binEval_84154e139590)
numFolds: number of folds for cross validation (>= 2) (default: 3, current: 5)
parallelism: the number of threads to use when running parallel algorithms (default: 1...


In [81]:
rfCvModel.explainParams

estimator: estimator for selection (current: rfc_3d7f3034b832)
estimatorParamMaps: param maps for the estimator (current: [Lorg.apache.spark.ml.param.ParamMap;@8e1daef)
evaluator: evaluator used to select hyper-parameters that maximize the validated metric (current: binEval_2183c3d46ce3)
numFolds: number of folds for cross validation (>= 2) (default: 3, current: 3)
seed: random seed (default: -1191137437)


Reference:
https://hub.packtpub.com/how-to-win-kaggle-competition-with-apache-sparkml/

In [80]:
// import org.apache.spark.ml.feature.OneHotEncoder
// import org.apache.spark.ml.feature.{OneHotEncoderEstimator, OneHotEncoderModel}

// // val indexColumns = Array("SEX", "EDUCATION", "MARRIAGE", "AGE", "PAY_1", "PAY_2", "PAY_3", "PAY_4", "PAY_5", "PAY_6")
// val indexColumns = Array("SEX", "EDUCATION", "MARRIAGE", "AGE")

// val encoder = new OneHotEncoderEstimator()
//   .setInputCols(indexColumns)
//   .setOutputCols(indexColumns map (name => s"${name}_vec"))

// val cc_enc = encoder.fit(cc).transform(cc)

Name: Syntax Error.
Message: 
StackTrace: 

In [81]:
// import org.apache.spark.ml.feature.VectorAssembler 
// import org.apache.spark.ml.linalg.Vectors

// val featureColumns = Array("SEX_vec", "EDUCATION_vec", "MARRIAGE_vec", "AGE_vec", "LIMIT_BAL", "PAY_1", "PAY_2", "PAY_3", "PAY_4", "PAY_5", "PAY_6", "BILL_AMT1", "BILL_AMT2", "BILL_AMT3", "BILL_AMT4", "BILL_AMT5", "BILL_AMT6", "PAY_AMT1", "PAY_AMT2", "PAY_AMT3", "PAY_AMT4", "PAY_AMT5", "PAY_AMT6")
// // val featureColumns = cc.columns.filter(_ !="label").filter(_ !="ID")

// var va = new VectorAssembler()
//     .setInputCols(featureColumns)
//     .setOutputCol("features")

// val cc_feat = va.transform(encoder.fit(cc).transform(cc))

Name: Syntax Error.
Message: 
StackTrace: 

In [82]:
// import org.apache.spark.ml.feature.Normalizer
// import org.apache.spark.ml.linalg.Vectors

// // Normalize each Vector using $L^1$ norm.
// val normalizer = new Normalizer()
//   .setInputCol("features")
//   .setOutputCol("normFeatures")
//   .setP(1.0)

// val l1NormData = normalizer.transform(cc_feat)
// println("Normalized using L^1 norm")
// l1NormData.show()

// // Normalize each Vector using $L^\infty$ norm.
// val lInfNormData = normalizer.transform(cc_feat, normalizer.p -> Double.PositiveInfinity)
// println("Normalized using L^inf norm")
// lInfNormData.show()

Name: Syntax Error.
Message: 
StackTrace: 

In [83]:
import org.apache.spark.ml.{Pipeline, PipelineModel}

//Create an array out of individual pipeline stages 
// var transformers = Array(encoder,va)
// var transformers = Array(va,normalizer)
var transformers = Array(scalerVectorAssembler, ccScaler, hot_encoder, finalVectorAssembler)

var pipeline = new Pipeline().setStages(transformers).fit(cc)

var transformed = pipeline.transform(cc)

transformers = Array(vecAssembler_daa01bb807f7, stdScal_7ca8cdaf8f2f, oneHotEncoder_f7081504b73c, vecAssembler_373cd1ff51a4)
pipeline = pipeline_5b610c7ea989
transformed = [ID: int, LIMIT_BAL: int ... 29 more fields]


[ID: int, LIMIT_BAL: int ... 29 more fields]

In [84]:
import org.apache.spark.ml.classification.RandomForestClassifier 

var rf = new RandomForestClassifier()
    .setLabelCol("DEFAULT")
    .setFeaturesCol("features")

var model = new Pipeline().setStages(transformers :+ rf).fit(cc) 
var result = model.transform(cc)

rf = rfc_848dcc5fb539
model = pipeline_e41369cfdd47
result = [ID: int, LIMIT_BAL: int ... 32 more fields]


[ID: int, LIMIT_BAL: int ... 32 more fields]

In [85]:
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator 

val evaluator = new BinaryClassificationEvaluator()
    .setLabelCol("DEFAULT")

import org.apache.spark.ml.param.ParamMap

var evaluatorParamMap = ParamMap(evaluator.metricName -> "areaUnderROC") 
var aucTraining = evaluator.evaluate(result, evaluatorParamMap)

evaluator = binEval_e93d0ac69aaf
evaluatorParamMap = 
aucTraining = 0.7762125622496241


{
	binEval_e93d0ac69aaf-metricName: areaUnderROC
}


0.7762125622496241

In [ ]:
import org.apache.spark.ml.tuning.{CrossValidator, ParamGridBuilder} 

var paramGrid = new ParamGridBuilder()
    .addGrid(rf.numTrees, 3 :: 5 :: 10 :: 30 :: 50 :: 70 :: 100 :: 150 :: Nil)
    .addGrid(rf.featureSubsetStrategy, "auto" :: "all" :: "sqrt" :: "log2" :: "onethird" :: Nil)
    .addGrid(rf.impurity, "gini" :: "entropy" :: Nil)
    .addGrid(rf.maxBins, 2 :: 5 :: 10 :: 15 :: 20 :: 25 :: 30 :: Nil)
    .addGrid(rf.maxDepth, 3 :: 5 :: 10 :: 15 :: 20 :: 25 :: 30 :: Nil)
    .build()

var crossValidator = new CrossValidator()
    .setEstimator(new Pipeline().setStages(transformers :+ rf))
    .setEstimatorParamMaps(paramGrid)
    .setNumFolds(5)
    .setEvaluator(evaluator)

var crossValidatorModel = crossValidator.fit(cc)

var newPredictions = crossValidatorModel.transform(cc)

In [ ]:
var bestPipelineModel = crossValidatorModel.bestModel.asInstanceOf[PipelineModel]

var stages = bestPipelineModel.stages

import org.apache.spark.ml.classification.RandomForestClassificationModel 

val rfStage = stages(stages.length-1).asInstanceOf[RandomForestClassificationModel] 

rfStage.getNumTrees
rfStage.getFeatureSubsetStrategy
rfStage.getImpurity
rfStage.getMaxBins
rfStage.getMaxDepth